src/type_checker.cpp

```diff
@@ -140,6 +140,51 @@ void TypeChecker::Visit(ContinueStmtNode& continue_stmt) {
   }
 }
 
+namespace {
+/// @brief A shared state to convey the presence of a default label in a switch
+/// statement.
+/// @note To allow nested switch statements, the state is stacked.
+auto
+    switch_already_has_default  // NOLINT(cppcoreguidelines-avoid-non-const-global-variables)
+    = std::vector<bool>{};
+}  // namespace
+
+void TypeChecker::Visit(SwitchStmtNode& switch_stmt) {
+  switch_stmt.ctrl->Accept(*this);
+  if (switch_stmt.ctrl->type != ExprType::kInt) {
+    // TODO: statement requires expression of integer type
+  }
+  body_types.push_back(BodyType::kSwitch);
+  switch_already_has_default.push_back(false);
+  switch_stmt.stmt->Accept(*this);
+  switch_already_has_default.pop_back();
+  body_types.pop_back();
+  // TODO: No two of the case constant expressions in the same switch statement
+  // shall have the same value (we need constant expression support on this).
+}
+
+void TypeChecker::Visit(CaseStmtNode& case_stmt) {
+  if (!IsInBodyOf(BodyType::kSwitch)) {
+    // TODO: 'case' statement not in switch statement
+  }
+  case_stmt.expr->Accept(*this);
+  if (case_stmt.expr->type != ExprType::kInt) {
+    // TODO: expression is not an integer constant expression
+  }
+  case_stmt.stmt->Accept(*this);
+}
+
+void TypeChecker::Visit(DefaultStmtNode& default_stmt) {
+  if (!IsInBodyOf(BodyType::kSwitch)) {
+    // TODO: 'default' statement not in switch statement
+  }
+  assert(!switch_already_has_default.empty());
```

Comment:
I think the logic here is a bit confusing because it feels unintuitive at first. I would think that `switch_already_has_default` should be empty. Then I checked line 158, realizing that it is correct.

Reply:
I cannot think of a better solution in the moment, maybe we just leave it.

Reply:
Maybe we add a comment to clarify that this sharing state is only meaningful for `switch`es?
Since the AST tree is quite general, I couldn't find a way to not let other nodes access the `switch_already_has_default`.